### ISSUE TITLE
Class-related methods fail to meet w3c standard for whitespace handling

### ISSUE THREAD

**Issue by zzzzBov** [jquery/jquery#3072]
[I opened a related discussion about whitespace on WICG][1] that may be worth a read before any changes are necessarily made regarding this.

the `*Class` methods have consistently failed to handle ["space characters"][5] correctly, and I'd reported [a related bug][2] a few years ago.

There are exactly 5 characters that are used to separate classes in HTML/CSS. Those characters in ascending unicode order are:

* tab (`\t`, U+0009)
* line feed (`\n`, U+000A)
* form feed (`\f`, U+000C)
* carriage return (`\r`, U+000D)
* space (` `, U+0020)

When other characters that are matched by the regular expression pattern `\s` are used in classes, jQuery fails to behave correctly according to the spec.

A quick example of this failure:

```
$({
  nodeType: 1,
  getAttribute: function () {
    return 'foo'
  },
  setAttribute: function (name, val) {
    console.assert(val === 'foo foo\u00A0bar', 'no-break space should not be treated as a class separator');
    console.log(JSON.stringify(val));
  }
}).addClass('foo\u00A0bar');
```

which produces an assertion failure and `'foo bar'` logged to console.

[fiddle][3]

The problem is that [`classes.js` uses `rnotwhite`][4] which is (`/\S+/g`) and *not* an appropriate regular expression to use for splitting strings of classes.

[1]: https://discourse.wicg.io/t/whitespace-is-hard-and-buggy-can-we-normalize-it/1436/1
[2]: https://bugs.jquery.com/ticket/13835
[3]: https://jsfiddle.net/8960sop4/
[4]: https://github.com/jquery/jquery/blob/master/src/attributes/classes.js#L26
[5]: https://www.w3.org/TR/html5/infrastructure.html#space-character

**Comment by dmethvin** [comment]
So it's just a case of using that specific list (negated) as our regexp, right? Is there any problem with updating that now? Per the other discussion I understand that some standards define whitespace differently but for the purpose of fixing our class methods it doesn't matter.

**Comment by gibson042** [comment]
We use `rnotwhite` for a few purposes beyond parsing a list of classes:
* [removing multiple attributes](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/src/attributes/attr.js#L90)
* [removing multiple data properties](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/data/Data.js#L128)
* [parsing Callbacks flags](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/src/callbacks.js#L9)
* [adding](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/event.js#L141)/[removing](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/event.js#L223) multiple events
* [specifying multiple animatable properties](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/effects.js#L414)
* [separating elements in ajax dataTypes](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/ajax.js#L532) [processing chains](https://github.com/jquery/jquery/blob/91850ecbbe04ad8f5d89dc050aa1d9002047c435/src/ajax.js#L65)

I'm fine with adopting the HTML/CSS whitespace definitions for all of them and updating `rnotwhite` accordingly.

**Comment by zzzzBov** [comment]
Whatever is chosen, be sure to add unit tests so that these issues don't creep in again later.

if the fix is just to change [`rnotwhite`][1] to `/[^\t\n\f\r ]+/g` it should be sufficient to fix the issues with classes, however it's possible that it may result in unicode errors for any usages where other unicode space characters were used as separating characters.

[1]: White_Space characters

**Comment by dmethvin** [comment]
@zzzzBov did you want to create a pull request for this?  It sounded possible that we might need separate patterns if there are different definitions of whitespace in different specs, but that would be something worth researching as part of the PR.

**Comment by zzzzBov** [comment]
@dmethvin 

> did you want to create a pull request for this?

Not really, I'd actually rather than the HTML/CSS specs be updated so that `\s` is the appropriate selector for whitespace because it would make more sense. Notice that literally no one else using jQuery ran into issues with this behavior and reported it, and the only reason I ran into it is because I went looking for it.

If jQuery could use `/\s/` and `/\S/` it would have significant data savings given how many times it's downloaded daily.

**Comment by dmethvin** [comment]
Okay, so should this issue be closed? We'd only leave it open if there is something we need to do in jQuery.

**Comment by zzzzBov** [comment]
@dmethvin I don't know what the right answer is. I can't decide for the jQuery team whether it's appropriate to diverge from what is explicitly specified behavior, or whether it's more important to be accurate.

> We'd only leave it open if there is something we need to do in jQuery.

Either way a change needs to be made. Either `/\s/` and `/\S/` are used or `/[ \r\n\f\t]/` and `/[^ \r\n\f\t]/` are used; currently usage is mixed, so someone needs to pick one and run with it.

**Comment by gibson042** [comment]
> I'd actually rather than the HTML/CSS specs be updated so that `\s` is the appropriate selector for whitespace

There's approximately zero chance of that happening, since `\s` [incorporates by reference](http://www.ecma-international.org/ecma-262/6.0/#sec-white-space) the Unicode "Separator, space" (Zs) [General Category](http://unicode.org/reports/tr44/#General_Category_Values), which is non-static (as opposed to the permanently fixed set of space/tab/line feed/form feed/carriage return used by both [HTML](https://html.spec.whatwg.org/multipage/infrastructure.html#space-character) and [CSS](https://www.w3.org/TR/2011/REC-css3-selectors-20110929/#whitespace)).

And since there _is_ cross-spec agreement on what whitespace separates tokens, I think we should follow suit as mentioned [above](https://github.com/jquery/jquery/issues/3072#issuecomment-213570315). We already use `[\x20\t\r\n\f]` or similar in six regular expression patterns (including [`rclass`](https://github.com/jquery/jquery/blob/305f193aa57014dc7d8fa0739a3fefd47166cd44/src/attributes/classes.js#L10)) to match those specs; updating `rnotwhite` won't change our gzipped size by more than a byte or two, and is actually slightly more likely to _reduce_ size than increase it. At the expense of `.removeData("foo\u1680bar")` removing two data properties, which I contend is a bug anyway since all our documentation specifies "space-separated".

**Comment by zzzzBov** [comment]
> There's approximately zero chance of that happening, since `\s` incorporates by reference the Unicode "Separator, space" (Zs) General Category, which is non-static...

[I'm pretty certain the HTML and CSS specs can handle it][1].

> And since there is cross-spec agreement on what whitespace separates tokens, I think we should follow suit as mentioned above.

I can't and won't disagree with this approach, I still maintain that [it would be convenient to normalize whitespace][2]

[1]: https://www.w3.org/TR/html5/single-page.html#white_space
[2]: https://discourse.wicg.io/t/whitespace-is-hard-and-buggy-can-we-normalize-it/1436

**Comment by gibson042** [comment]
Related: gh-3003.

**Comment by mgol** [comment]
Sounds like a good candidate for 3.0.1.
